source/sliding_puzzle.py

```diff
@@ -124,7 +139,67 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        random.shuffle(self.rows)
+        # empty one of the rows
+        random.choice(random.choice(self.rows)).clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece._data == []
+        ][0]
+        return PiecePosition(x=x, y=y, index=index)
+
+    def move_up(self) -> None:
+        """Move the piece above the empty piece down"""
+        empty_pos = self._get_empty_piece_position()
+        # return if empty piece is on the first row
+        top_row = self.rows[0]
+        if empty_pos.index in [piece.index for piece in top_row]:
+            return
```

Comment:
```suggestion
        if empty_pos.x == 0:
            return  # We cannot move the empty piece up
```

Reply:
Same as the similar changes you requested, 0 is not the only value that indicate that the empty piece can not be moved up, at all position on the first row we cannot move up

Reply:
But that would have different y values? All positions on the first row has `x == 0`